internal/v1alpha1/service_binding.go

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// ServiceBindingApplicationReference defines a subset of corev1.ObjectReference with extensions
+type ServiceBindingApplicationReference struct {
+	ObjectReference `json:",inline"`
+	// Containers describes which containers in a Pod should be bound to
+	Containers []intstr.IntOrString `json:"containers,omitempty"`
+	// Selector is a query that selects the application or applications to bind the service to
+	Selector metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// ServiceBindingServiceReference defines a subset of corev1.ObjectReference
+type ServiceBindingServiceReference struct {
+	ObjectReference `json:",inline"`
+}
+
+// ObjectReference is a subset of corev1.ObjectReference
+type ObjectReference struct {
+	// API version of the referent.
+	APIVersion string `json:"apiVersion,omitempty"`
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	Kind string `json:"kind,omitempty"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name,omitempty"`
+}
+
+// ServiceBindingEnvVar defines a mapping from the value of a Secret entry to an environment variable
+type ServiceBindingEnvVar struct {
+	// Name is the name of the environment variable
+	Name string `json:"name"`
+	// Key is the key in the Secret that will be exposed
+	Key string `json:"key"`
+}
+
+// ServiceBindingMapping defines a mapping from the existing collection of Secret values to a new Secret entry.
+type ServiceBindingMapping struct {
+	// Name is the name of the mapped Secret entry
+	Name string `json:"name"`
+	// Value is the value of the new Secret entry.  Contents may be a Go template and refer to the other secret entries
+	// by name.
+	Value string `json:"value"`
+}
+
+// ServiceBindingSpec defines the desired state of ServiceBinding
+type ServiceBindingSpec struct {
+	// Name is the name of the service as projected into the application container.  Defaults to .metadata.name.
+	Name string `json:"name,omitempty"`
+	// Type is the type of the service as projected into the application container
+	Type string `json:"type,omitempty"`
+	// Provider is the provider of the service as projected into the application container
+	Provider string `json:"provider,omitempty"`
+	// Application is a reference to an object that fulfills the PodSpec duck type
+	Application ServiceBindingApplicationReference `json:"application"`
+	// Service is a reference to an object that fulfills the ProvisionedService duck type
+	Service ServiceBindingServiceReference `json:"service"`
+	// EnvVars is the collection of mappings from Secret entries to environment variables
+	EnvVars []ServiceBindingEnvVar `json:"env,omitempty"`
```

Comment:
Perhaps we should add the `+optional` as added here for optional fields https://github.com/k8s-service-bindings/spec/pull/64/files#diff-448ba65aaf058ec1a8c0d15e8960acccR118

Reply:
I'm in favor of this for clarity, but it won't actually change the schema as only `application` and `service` are [currently required](https://github.com/k8s-service-bindings/spec/pull/64/files#diff-390050d517cb57200f1a0d5b6dc688fbR181-R183).